### PR TITLE
8261268: LOAD_INSTANCE placeholders unneeded for parallelCapable class loaders

### PR DESCRIPTION
**Pull request by coleenp**
See CR for more details.  This optimizes some code for class loading.
Tested with tier1-6.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8261268](https://bugs.openjdk.java.net/browse/JDK-8261268): LOAD_INSTANCE placeholders unneeded for parallelCapable class loaders

### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to 9eb1634b04672d6f682f6066bf41c42aabfb3011
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2469/head:pull/2469`
`$ git checkout pull/2469`

### CODE REVIEW

File: src/hotspot/share/classfile/placeholders.cpp [hunk 1]
```diff
@@ -104,14 +104,17 @@ void PlaceholderEntry::set_threadQ(SeenThread* seenthread, PlaceholderTable::cla
 
 // Doubly-linked list of Threads per action for class/classloader pair
 // Class circularity support: links in thread before loading superclass
-// bootstrapsearchpath support: links in a thread before load_instance_class
+// bootstrap loader support:  links in a thread before load_instance_class
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/placeholders.cpp:107]: I don't think the loader is what the comment is referring to. The reference to bootstrapsearchpath relates to this comment in SD::resolve_instance_class_or_null:
    // add placeholder entry to record loading instance class
    // Five cases:
    // All cases need to prevent modifying bootclasssearchpath
    // in parallel with a classload of same classname

File: src/hotspot/share/classfile/systemDictionary.cpp [hunk 2]
```diff
@@ -584,14 +581,7 @@ InstanceKlass* SystemDictionary::handle_parallel_super_load(
...
   unsigned int name_hash = dictionary->compute_hash(name);
 
-  // superk is not used, resolve_super called for circularity check only
-  // This code is reached in two situations. One if this thread
-  // is loading the same class twice (e.g. ClassCircularity, or
-  // java.lang.instrument).
-  // The second is if another thread started the resolve_super first
-  // and has not yet finished.
-  // In both cases the original caller will clean up the placeholder
-  // entry on error.
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/systemDictionary.cpp:594]: Is this commentary really no longer applicable? While I don't fully understand it it does indicate the general conditions under which we will call handle_parallel_super_load.

File: src/hotspot/share/classfile/systemDictionary.cpp [hunk 3]
```diff
@@ -1862,8 +1837,6 @@ void SystemDictionary::check_constraints(unsigned int name_hash,
       }
     }
 
-    DEBUG_ONLY(if (is_parallelCapable(class_loader)) verify_placeholder(name, loader_data));
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/systemDictionary.cpp:1865]: I can't help but feel that somewhere in the loading process there should be a more general assertion that, not only is there a placeholder present, but that it is for this thread. Though such a check would also need to check the kind of placeholder found. Future RFE perhaps.

**Review by dholmes-ora — APPROVED** [review]
Hi Coleen,

The actual logic changes seem sound (though some of the reorganisation was a little hard to follow).

I have some comments on comments, but overall this seems okay provided all testing passes.

Thanks,
David

File: src/hotspot/share/classfile/placeholders.cpp (hunk 1 above)
Comment (coleenp) [reply src/hotspot/share/classfile/placeholders.cpp:107]: I don't see any code that does this though.  The code used to compute_loader_lock_object() and return NULL in JVMTI where it was modifying the bootclass search path.  I wonder if that code should have a wait on the placeholder.  I'll investigate.

File: src/hotspot/share/classfile/systemDictionary.cpp (hunk 2 above)
Comment (coleenp) [reply src/hotspot/share/classfile/systemDictionary.cpp:594]: I'm in progress of investigating why this code exists.   The last two lines were false, and these lines seem to repeat the comments above handle_parallel_super_load.  Pointing out java.lang.instrument is even more befuddling.  There may have been a case once where two threads tried to load the same class in RedefineClasses.  I'm not sure why the LOAD_INSTANCE wouldn't have stopped them, just like it stops all classes from being reloaded.

File: src/hotspot/share/classfile/systemDictionary.cpp (hunk 3 above)
Comment (coleenp) [reply src/hotspot/share/classfile/systemDictionary.cpp:1865]: In this case, we call check_constraints() after loadClass so the placeholder doesn't exist for parallelCapable class loaders (we do want one for the define class path).  Not sure what sort of assertion would be helpful here.

File: src/hotspot/share/classfile/systemDictionary.cpp (hunk 2 above)
Comment (coleenp) [reply src/hotspot/share/classfile/systemDictionary.cpp:594]: The second sentence of this comment seems correct but is repeated above handle_parallel_super_load. I was going to restore some of this but it doesn't add any information that isn't in the comment above this function.

File: src/hotspot/share/classfile/placeholders.cpp (hunk 1 above)
Comment (coleenp) [reply src/hotspot/share/classfile/placeholders.cpp:107]: ClassLoader::load_class reads the bootstrapsearchpath lock free.  If it's modified after it's read, then the new entry isn't found.  I don't think there's a requirement for locking.

**Comment by coleenp** [comment]
>The actual logic changes seem sound (though some of the reorganisation was a little hard to follow).

Unfortunately the diffs are bad because I did two things.  I changed the indentation and I check for the class in the dictionary before adding the LOAD_INSTANCE placeholder for the !parallelCapable() && !bootclass case.

> I have some comments on comments, but overall this seems okay provided all testing passes.
Yup, thanks!

**Review by iklam — APPROVED** [review]
LGTM

**Comment by coleenp** [comment]
Thanks Ioi.  I reworded the comment we talked about offline.
+    //    These class loaders lock a per-class object lock when ClassLoader.loadClass()
+    //    is called. A LOAD_INSTANCE placeholder isn't used for mutual exclusion.

**Review by iklam — APPROVED** [review]

**Comment by coleenp** [comment]
@dholmes-ora thanks for double checking the older code.
/integrate
